### PR TITLE
Reimplement simulator UI in React

### PR DESCRIPTION
**Pull request by xian**
Use façades between domain objects and React.

### CODE REVIEW

File: src/jsMain/kotlin/baaahs/sim/ui/NetworkPanel.kt [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+    override fun componentWillUpdate(nextProps: Props, nextState: State) {
+        props.simulator?.network?.removeObserver(this)
+        nextProps.simulator?.network?.addObserver(this)
+    }
+
+    override fun notifyChanged() {
+        if (!changedSinceRender) {
+            changedSinceRender = true
+            setState({ s -> s }) {
```
Comment (xian) [inline src/jsMain/kotlin/baaahs/sim/ui/NetworkPanel.kt:40]: `setState()` here always causes a _synchronous_ call to render, so _every single packet_ causes a full render, which is super slow.

**Comment by xian** [comment]
@r3dcrosse / @hassoncs whutcha think? Could be optimized a bit more for fewer renders but a big improvement I think.

File: src/jsMain/kotlin/baaahs/JsMain.kt [hunk 2]
```diff
@@ -14,17 +13,7 @@ import kotlin.browser.window
...
-
-        fun <T> nuffin(): T {
-            @Suppress("UNCHECKED_CAST")
-            return null as T
-        }
-        GadgetDisplay(nuffin()) {}
-
-        GlslPreview(nuffin(), nuffin(), null)
-    }
+    if (1 + 1 == 3) dce()
```
Comment (r3dcrosse) [inline src/jsMain/kotlin/baaahs/JsMain.kt:16]: This still trips me up when I see this line 😄

File: src/jsMain/kotlin/baaahs/jsImpl.kt [hunk 3]
```diff
@@ -42,7 +42,7 @@ private fun logMessage(level: String, message: String, exception: Throwable?) {
         "ERROR" -> console.error(message, exception)
         "WARN" -> console.warn(message, exception)
         "INFO" -> console.info(message, exception)
-        "DEBUG" -> console.log(message, exception)
+        "DEBUG" -> console.asDynamic().debug(message, exception)
```
Comment (r3dcrosse) [inline src/jsMain/kotlin/baaahs/jsImpl.kt:45]: I don't understand why the kotlin stdlib doesn't support `console.debug`...

https://github.com/JetBrains/kotlin/blob/master/libraries/stdlib/js/src/kotlin/debug.kt

It's totally a thing! https://developer.mozilla.org/en-US/docs/Web/API/Console/debug

File: src/jsMain/kotlin/baaahs/ui/ReactIsCrap.kt [hunk 4]
```diff
@@ -0,0 +1,7 @@
+@file:JsModule("react-dom")
+
+package baaahs.ui
+
+// See https://reactjs.org/docs/react-dom.html
+
+external fun unstable_batchedUpdates(callback: dynamic)
```
Comment (r3dcrosse) [inline src/jsMain/kotlin/baaahs/ui/ReactIsCrap.kt:7]: ಠ_ಠ

In general, I don't recommend using any of the react apis that preface with `unstable_`, but it seems to work right now and I don't know of any alternatives so ¯\_(ツ)_/¯

File: src/jsMain/kotlin/baaahs/sim/ui/PinkyPanel.kt [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+                val beat = beatData.beatWithinMeasure(pinky.clock).toInt()
+                val bpm = beatData.bpm
+                val beatConfidence = beatData.confidence
+
+                b { +"Beats: " }
+                span { // beatConfidenceElement
+                    +"[confidence: ${beatConfidence.percent()}]"
+                }
+                br { }
+                styledDiv { +"1"; if (beat == 0) css { +"selected" } }
```
Comment (r3dcrosse) [inline src/jsMain/kotlin/baaahs/sim/ui/PinkyPanel.kt:88]: This syntax is really interesting and I don't know whether to be appalled or maybe you're onto something here?

Are there any links to documentation/discussion threads about what looks like sass but kinda like jsx but totally kotlin? 

I still will ask: what is wrong with using jsx?

File: src/jsMain/kotlin/baaahs/JsMain.kt (hunk 2 above)
Comment (xian) [reply src/jsMain/kotlin/baaahs/JsMain.kt:16]: Yeah, it's ugly and I'm certain there's a better way to do this, I'm just [too lazy](https://kotlinlang.org/docs/reference/javascript-dce.html#excluding-declarations-from-dce) at the moment to find it.

File: src/jsMain/kotlin/baaahs/jsImpl.kt (hunk 3 above)
Comment (xian) [reply src/jsMain/kotlin/baaahs/jsImpl.kt:45]: Yep, agreed. :-p

File: src/jsMain/kotlin/baaahs/ui/ReactIsCrap.kt (hunk 4 above)
Comment (xian) [reply src/jsMain/kotlin/baaahs/ui/ReactIsCrap.kt:7]: Bums me out too. This should really be a supported part of the library.

File: src/jsMain/kotlin/baaahs/sim/ui/PinkyPanel.kt (hunk 5 above)
Comment (xian) [reply src/jsMain/kotlin/baaahs/sim/ui/PinkyPanel.kt:88]: Heheh yeah I'm not used to it yet either.

`styledDiv` is gross; it's the same as `div` with access to CSS config.

I'm using Kotlin for the React rendering here cuz I really like having edit-time type checking, autocomplete, and refactoring. It compiles into CSS just like JSX, so, same same.

Comment (xian) [reply src/jsMain/kotlin/baaahs/sim/ui/PinkyPanel.kt:88]: (and JSX still has the bewildering panoply of JS idioms and historical brokenness (`function(){}.bind(this)` et al.), which Kotlin fixes.)

Comment (xian) [reply src/jsMain/kotlin/baaahs/sim/ui/PinkyPanel.kt:88]: Fine with me if you wanna use JSX here though, this is just easier/more familiar to me.

Comment (xian) [reply src/jsMain/kotlin/baaahs/sim/ui/PinkyPanel.kt:88]: But to answer your question, it's equivalent to:

```html
<div class={beat == 0 ? "selected" : ""}>1</div>
```

Some Kotlin DSLs use the `+` operator as a sort of "contextually add this thing" command. So inside a tag it's _add string_; inside a CSS block it's _add class_.

**Pull request merged**